### PR TITLE
Issue/v4/1638 object categories

### LINKED ISSUE #1638: Object Categories data model + API
Implement object categories 
see https://github.com/orgs/bedita/teams/core/discussions/11

### PR DESCRIPTION
**Pull request by stefanorosanelli**
This PR fixes #1638 

## Tables 

• new `categories` and  `object_categories` tables defined

## Model 

• new model tables: `CategoriesTable`, `ObjectCategoryTable`, `TagsTable`, `ObjectTagsTable`
• new model entities: `Category`, `Tag`, `ObjectCategory`
• new helper `CategoriesBehavior` to save `categories` and `tags` on `ObjectsTable`
• `InheritanceEventHandler` has been modified in order to avoid `save` actions on tables extending `Objects` when `tags` or `categories` array are set in entity data: those actions MUST be performed in `ObjectsTable` only using `CategoriesBehavior`. A new `$excludeDescendantsSave` array has been defined, but it's of course a temporary solution.
 
### Modeling API

This version provides some basic endpoints, but filters and other refinements are needed, in a future issue / PR
Now we have:
• create & index: `POST/GET /model/{object_type}/categories`
• read, modify & remove: `GET/PATCH/DELETE /model/{object_type}/categories/{id}`

### Requests and responses

To create or update and object with categories/tags you have to simply add to `attributes` data body an array like this for tags:
```json
                    "tags": [
                        {"name": "first-tag"}
                    ],
```

and like this for categories:
```json
                    "categories": [
                        {"name": "first-category"},
                        {"name": "second-category"},
                    ],
```

Request response will contain also `label` and `params` like:
```json
                    "categories": [
                        {  
                             "name": "first-category",
                             "label": "Firs category",
                             "params": "100"
                         }
                    ],
```
Where:
 * `label` is a human readable category name
 * `params` may contain any JSON but it's not writable at the moment

### Filters 

Use categories or tags filter with comma separated list of categories or tags names like:

`GET /documents?filter[categories]=one,two`

### CODE REVIEW

**Comment by batopa** [comment]
@stefanorosanelli there are some conflicts, please can you resolve them?

**Comment by stefanorosanelli** [comment]
> @stefanorosanelli there are some conflicts, please can you resolve them?

Should be ok now

**Review by didoda — APPROVED** [review]

File: plugins/BEdita/API/tests/IntegrationTest/TagsCategoriesTest.php [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    public function testUpdate(array $expected, string $type, string $id, array $attributes)
+    {
+        // Patch object
+        $data = compact('type', 'id', 'attributes');
+        $this->configRequestHeaders('PATCH', $this->getUserAuthHeader());
+        $this->patch("/$type/$id", json_encode(compact('data')));
+        $this->assertResponseCode(200);
+        $this->assertContentType('application/vnd.api+json');
+        $result = json_decode((string)$this->_response->getBody(), true);
+// pr($result);
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/TagsCategoriesTest.php]: remove comment

File: plugins/BEdita/API/tests/IntegrationTest/TagsCategoriesTest.php [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+        // View object
+        $this->configRequestHeaders();
+        $this->get("/$type/$id");
+        $result = json_decode((string)$this->_response->getBody(), true);
+        $this->assertResponseCode(200);
+        $this->assertContentType('application/vnd.api+json');
+
+// pr($result);
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/TagsCategoriesTest.php]: remove comment

File: plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                ],
+                [
+                    'tags' => [
+                        ['name' => 'first-tag'],
+                    ]
+                ],
+                2,
+                'Documents',
+            ],
+
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php]: remove empty line here

File: plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                ],
+                [
+                    'categories' => [
+                        ['name' => 'second-cat'],
+                    ]
+                ],
+                4,
+                'Profiles',
+            ],
+
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php]: remove empty line here

File: plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                    ]
+                ],
+                4,
+                'Profiles',
+            ],
+        ];
+    }
+
+    /**
+     * Test correct save of custom properties.
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php]: The comment here is wrong... I suppose :smile:

File: plugins/BEdita/Core/tests/TestCase/Model/Table/ObjectCategoriesTableTest.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    /**
+     * Test initialize method
+     *
+     * @return void
+     */
+    public function testInitialize()
+    {
+        $this->markTestIncomplete('Not implemented yet.');
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Table/ObjectCategoriesTableTest.php]: Why does it mark as incomplete? We can implement a basic test here. Same for others test methods.

File: plugins/BEdita/Core/src/Model/Behavior/CategoriesBehavior.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    protected function prepareData(string $item, EntityInterface $entity)
+    {
+        if (!$entity->isDirty($item)) {
+            $entity->unsetProperty($item);
+
+            return;
+        }
+        // Check if `Tags` or `Categories` associations are enabled
+        $objectType = $this->getTable()
+                    ->getAssociation('ObjectTypes')
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Behavior/CategoriesBehavior.php]: too much indentation here

**Review by batopa — CHANGES_REQUESTED** [review]
I think a better implementation could be done having two Tables and Entities i.e. `CategoriesTable`, `TagsTable`, `Category` and `Tag`.

Every table could use a common `CategoriesTrait` that implements common functions while the single tables should specify their peculiarity.

For example for `TagsTable` we can avoid to have

```php
$this->belongsTo('ObjectTypes', [
    'foreignKey' => 'object_type_id',
    'className' => 'BEdita/Core.ObjectTypes'
]);
$this->belongsTo('ParentCategories', [
    'className' => 'BEdita/Core.Categories',
    'foreignKey' => 'parent_id'
]);
$this->hasMany('ChildCategories', [
    'className' => 'BEdita/Core.Categories',
    'foreignKey' => 'parent_id'
]);
$this->addBehavior('Tree', [
    'left' => 'tree_left',
    'right' => 'tree_right',
]);
```

Every Table should implement a `beforeFind()`method to add the condition on `object_type_id` (NULL OR NOT NULL).

`findEnabledTags()` and `findEnabledCategories()` could be simplified as

```
protected function findEnabled(Query $query)
{
    return $query->where([
        $this->aliasField('enabled') => true,
    ]);
}
```

And also `findCategoriesIds()` and `findTagsIds()` can be simplified and having the same name.

The `Tag` and `Category` entities will be equal but we could easily customize them in the future.

In this way I think its more clear to use tag and categories. For example

```php
// every time I want tags I know I have to use TagsTable 
// instead of use CategoriesTable with specific conditions

$categories = TableRegistry::getTableLocator()->get('Categories')->find('enabled');
$tags = TableRegistry::getTableLocator()->get('Tags')->find('enabled');
```

The associations in `ObjectsTable` will change slighlty

```php
$this->belongsToMany('Categories', [
    'className' => 'BEdita/Core.Categories',
    'through' => 'BEdita/Core.ObjectCategories',
    'foreignKey' => 'object_id',
    'targetForeignKey' => 'category_id',
    'finder' => 'enabled',
    'cascadeCallbacks' => true,
]);
$this->belongsToMany('Tags', [
    'className' => 'BEdita/Core.Tags',
    'through' => 'BEdita/Core.ObjectCategories',
    'foreignKey' => 'object_id',
    'targetForeignKey' => 'category_id',
    'finder' => 'enabled',
    'cascadeCallbacks' => true,
]);
```

and some changes will have to done in `CategoriesBehavior`.

What do you think about it? Do you think it's worth it?

File: plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php (hunk 5 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/tests/TestCase/Model/Behavior/CategoriesBehaviorTest.php]: Probably :smiley:

File: plugins/BEdita/Core/tests/TestCase/Model/Table/ObjectCategoriesTableTest.php (hunk 6 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/tests/TestCase/Model/Table/ObjectCategoriesTableTest.php]: That's true, however for similar table classes like `ObjectPermissionsTable` and `ObjectPropertiesTable` test cases are similar, with incomplete tests... 
I'll try to add one here

Comment (stefanorosanelli) [reply plugins/BEdita/Core/tests/TestCase/Model/Table/ObjectCategoriesTableTest.php]: done

**Comment by stefanorosanelli** [comment]
> I think a better implementation could be done having two Tables and Entities i.e. `CategoriesTable`, `TagsTable`, `Category` and `Tag`.

I agree, didn't think it this way because we normally have a single table class for a db `resource` type table. But it makes sense, especially for future uses of the table classes.

I'll try to implement this refactor, if too problematic will decide how to proceed.

**Review by batopa — APPROVED** [review]

**Pull request merged**